### PR TITLE
Slow query logging

### PR DESCRIPTION
**Pull request by jschlather**
Builds in support for logging slow queries on the client side.

https://git.hubteam.com/HubSpot/sre-team/issues/99

### CODE REVIEW

File: java/jdbc/src/test/java/io/vitess/jdbc/ConnectionPropertiesTest.java [hunk 1]
```diff
@@ -174,8 +174,8 @@ public void testDriverPropertiesOutput() throws SQLException {
     assertEquals("grpcRetriesInitialBackoffMillis", infos[9].name);
     assertEquals("grpcRetriesMaxBackoffMillis", infos[10].name);
     assertEquals(Constants.Property.INCLUDED_FIELDS, infos[11].name);
-    assertEquals(Constants.Property.TABLET_TYPE, infos[23].name);
-    assertEquals(Constants.Property.TWOPC_ENABLED, infos[31].name);
+    assertEquals(Constants.Property.TABLET_TYPE, infos[24].name);
+    assertEquals(Constants.Property.TWOPC_ENABLED, infos[32].name);
```
Comment (acharis) [inline java/jdbc/src/test/java/io/vitess/jdbc/ConnectionPropertiesTest.java]: these seem unrelated. can we move the new constant around so that these indices don't need to change?

File: java/client/src/main/java/io/vitess/client/VTGateConnection.java [hunk 2]
```diff
@@ -112,10 +130,35 @@ public VTGateConnection(RpcClient client) {
                 }
               }, directExecutor()));
       vtSession.setLastCall(call);
+
+      if (slowQueryLoggingThresholdMillis != -1) {
+        call.addListener(new SlowQueryLogger(start, slowQueryLoggingThresholdMillis, query),
```
Comment (acharis) [inline java/client/src/main/java/io/vitess/client/VTGateConnection.java]: here and below you create a new SlowQueryLogger for each call to execute.
(i'm assuming that the plumbing of call.AddListener that i'm not seeing here will call `run` at the appropriate time)

did you profile this or anything to measure the overhead? this seems really heavyweight to me. is there any way around it so that we can reuse an object and just call into it?

File: java/client/src/main/java/io/vitess/client/VTGateConnection.java [hunk 3]
```diff
@@ -64,8 +70,18 @@
...
  */
 public class VTGateConnection implements Closeable {
-
+  private static final ExecutorService QUERY_LOGGING_EXECUTOR = new ThreadPoolExecutor(2, 2,
+          Long.MAX_VALUE,
+          TimeUnit.DAYS,
+          new LinkedBlockingQueue<>(50),
+          new ThreadFactoryBuilder()
+                  .setNameFormat("vitess-slow-query-logger-%d")
+                  .setDaemon(false)
```
[the remark this answers is not in the document]
Comment (jschlather) [reply java/client/src/main/java/io/vitess/client/VTGateConnection.java]: Nice catch, meant to make these daemon threads.

File: java/jdbc/src/test/java/io/vitess/jdbc/ConnectionPropertiesTest.java (hunk 1 above)
Comment (jschlather) [reply java/jdbc/src/test/java/io/vitess/jdbc/ConnectionPropertiesTest.java]: This test is way too brittle is my feeling. It's dependent on the order of the reflective access of the `ConnectionProperties` fields. I already declared the field last, so I'm not sure what else could be done.

File: java/client/src/main/java/io/vitess/client/VTGateConnection.java (hunk 2 above)
Comment (jschlather) [reply java/client/src/main/java/io/vitess/client/VTGateConnection.java]: I need to store references to the `start` and `query`, so that I can log them at the end. The listener API takes a `Runnable` and then calls `run` on it after the call has executed. I have to create a `Runnable` and it has to have those fields, in order for it to be useful. The execution is performed on an arbitrary thread, so I cannot store the values in a ThreadLocal.

Anyway, this is actually pretty light weight. It creates a single object, the size of which is much smaller than a ResultSet and similar in size to many of the objects created by GRPC along the way. The heaviest part here is really executing the runnable, but it's a very small statement and is a very low amount of work compared to everything the query does.

Comment (acharis) [reply java/client/src/main/java/io/vitess/client/VTGateConnection.java]: ok, sgtm

**Comment by acharis** [comment]
lgtm
